### ISSUE TITLE
Unable to use "git_checkout_index" to discard unstaged changes

### ISSUE THREAD

**Issue by aaronfranke** [libgit2/libgit2#5803]
I'm trying to implement a "Discard unstaged changes" button. This operation involves taking the content of the index (the staging area) and setting the files in the working tree to that, which means updating the working tree to match the content of the index. This is what the `git_checkout_index` method is described

```c
/**
 * Updates files in the working tree to match the content of the index.
 *
 * @param repo repository into which to check out (must be non-bare)
 * @param index index to be checked out (or NULL to use repository index)
 * @param opts specifies checkout options (may be NULL)
 * @return 0 on success, non-zero return value from `notify_cb`, or error
 *         code < 0 (use git_error_last for error details)
 */
GIT_EXTERN(int) git_checkout_index(
	git_repository *repo,
	git_index *index,
	const git_checkout_options *opts);
```

### Reproduction steps

I have a Git repo initialized in a `git_repository *` variable called `repo`. Since the second and third parameters are optional, I tried running the method with those set to NULL, like this:

```c
git_checkout_index(repo, NULL, NULL);
```

This method return 0 (which means no error was reported).

I am able to perform other operations fine, so I have libgit2 and my repo set up properly.

### Expected behavior

The expected behavior is that this method will update files in the working tree to match the content of the index, and therefore unstaged changes will be discarded.

### Actual behavior

Nothing happens. No files are modified. The method exits successfully but does nothing.

### Version of libgit2 (release number or SHA1)

0.28.4 (specifically 0.28.4+dfsg.1-2, this is the version that is included in Ubuntu 20.04)

### Operating system(s) tested

Only Linux (Ubuntu 20.04) was tested.

**Comment by ethomson** [comment]
I don't think it would affect the problem but I'm curious: are you speaking to libgit2 directly via C, using an existing binding to a language like .NET, or doing a custom FFI from another language?

**Comment by aaronfranke** [comment]
I'm using C directly. I am writing a program in C and I have libgit2 linked as a C library.

**Comment by ethomson** [comment]
Thanks - just curious based on your C# work on Godot.  I don't know offhand what the problem might be but I'll try to take a closer look this weekend.

**Comment by aaronfranke** [comment]
@ethomson I've been struggling to figure this out for a long time, but I have figured something out, maybe you can help me through it. This code below only discards *staged* files but keeps unstaged files, so it's the opposite of what I want.

```c
	// Get the index of the repository.
	git_index *index;
	git_repository_index(&index, repo);
	// Checkout HEAD with index as a baseline.
	git_checkout_options opt = GIT_CHECKOUT_OPTIONS_INIT;
	opt.baseline_index = index;
	git_checkout_head(repo, &opt);
```

From what I'm understanding, it seems that Git is only comparing HEAD and the baseline, so that Git will know which files need to be updated, and wherever they differ, writing what's in HEAD, and for all other files, ignore them.

So, in order to discard unstaged changes, we need to compare the index to the working tree, with the working tree as the baseline, so that Git will know which files need to be updated (and use `git_checkout_index` instead of the head version). 

However, I have absolutely no idea how to use the working directory as the baseline! `git_checkout_options` accepts either a `git_tree` or a `git_index` as the baseline. There doesn't seem to be any method to get the contents of the working tree as a `git_tree` or a `git_index`.

My assumption would be that if the baseline is unspecified, then it would treat the working tree as a baseline (it would make sense to me... the method is about updating the working tree, so why doesn't it use the working tree as a baseline?). But when I don't specify a baseline (or just pass `NULL` as the method description says is allowed), nothing happens with either method! Additionally, `git_checkout_head` says it "Updates the files in the index and the working tree", so surely there must be a way to specify the working tree as the baseline so that it can perform the documented behavior of "Updates the files in [...] the working tree".

**Comment by ethomson** [comment]
Hi @aaronfranke - thanks for the additional information, after reading this I think that I understand more clearly.  (Apologies that I did not, in fact, have time to take a look a few weekends ago.)

By default, checkout will preserve changes in the working directory.  (And if it cannot, it will raise a checkout conflict, preventing you from performing the checkout.)

You can see this on the git command line, for example:

```
% git checkout main
M	asdf.txt
Switched to branch 'main'
```

If you want to remove unstaged changes, you'll need to do the equivalent of a force checkout:

>  * - `GIT_CHECKOUT_FORCE` is at the opposite extreme, taking any action to
> *   make the working directory match the target (including potentially
> *   discarding modified files).

**Comment by aaronfranke** [comment]
@ethomson Thank you! This work as expected:

```c
	// Force-checkout from the index to discard unstaged changes.
	git_checkout_options opt = GIT_CHECKOUT_OPTIONS_INIT;
	opt.checkout_strategy = GIT_CHECKOUT_FORCE;
	git_checkout_index(repo, NULL, &opt);
```

Do you agree with me that either this behavior should be the default for `git_checkout_index`, or at least it should be documented and throw an error or warning, and especially don't encourage the user to pass `NULL` for the options? It doesn't make sense to me that a method for updating the working tree isn't allowed to update the working tree by default and silently fails to do so.

**Comment by extrawurst** [comment]
@aaronfranke this non destructive default behaviour is inline with regular `git checkout`

**Comment by aaronfranke** [comment]
@extrawurst Yes, you are correct. However, this method does not perform the same thing as `git checkout`.

What is the use case for this method without `GIT_CHECKOUT_FORCE` set? From what I can see, if `GIT_CHECKOUT_FORCE` is not specified, this method just does nothing at all.

**Comment by ethomson** [comment]
> What is the use case for this method without GIT_CHECKOUT_FORCE set? From what I can see, if GIT_CHECKOUT_FORCE is not specified, this method just does nothing at all.

I don't understand what you mean.  This is true if your goal is to overwrite unstaged changes.  Then, yes, you need to force.  This is precisely like git itself, isn't it?

If your goal is to stage some change and then update the working directory (for example, `git apply --index`) then this is precisely what you would do (and how our git_apply is implemented).
